File: Hard/n_queens_II/n_queens_II.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <cstdlib>
using namespace std;

class Solution {
    public:
        int totalNQueens(int n) {
            vector<int> queens(n, -1);
            int row = 0;
            int column = 0;
            int count = 0;

            while(row >= 0) {
                bool queenPlaced = false;
                while(column < n) {
                    bool safe = true;
                    for(int previousRow = 0; previousRow < row; previousRow++) {
                        bool sameColumn = queens[previousRow] == column;
                        bool sameDiagonal = abs(queens[previousRow] - column) == row - previousRow;
                        if(sameColumn || sameDiagonal) {
                            safe = false;
                            break;
                        }
                    }
                    if(safe) {
                        queens[row] = column;
                        if(row == n - 1) {
                            count++;
                            column++;
                        }
                        else {
                            row++;
                            column = 0;
                        }
                        queenPlaced = true;
                        break;
                    }
                    column++;
                }
                if(!queenPlaced && column == n) {
                    queens[row] = -1;
                    row--;
                    if(row >= 0) {
                        column = queens[row] + 1;
                        queens[row] = -1;
                    }
                }
            }
            return count;
        }
};

// Example usage:
int main() {
    Solution solution;
    int n1 = 4;
    int result1 = solution.totalNQueens(n1);
    cout << "Testing N-Queens II: " << endl;
    cout << "Input: n = " << n1 << endl; 
    cout << "Output: " << result1 << endl;

    cout << "" << endl;
    int n2 = 1;
    int result2 = solution.totalNQueens(n2);
    cout << "Input: n = " << n2 << endl;
    cout << "Output: " << result2 << endl;

    return 0;
}

File: Hard/n_queens_II/n_queens_II.py
from typing import List

def totalNQueens(n: int) -> int:
    columns = set()
    diagonal1 = set() # row - col
    diagonal2 = set() # row + col

    def backtrack(row):
        if row == n:
            return 1
        count = 0
        for col in range(n):
            if col in columns:
                continue
            if (row - col) in diagonal1:
                continue
            if (row + col) in diagonal2:
                continue
            columns.add(col)
            diagonal1.add(row - col)
            diagonal2.add(row + col)
            count += backtrack(row + 1)
            columns.remove(col)
            diagonal1.remove(row - col)
            diagonal2.remove(row + col)
        return count
    return backtrack(0)

# Example usage:
if __name__ == "__main__":
    print("Testing totalNQueens function:")
    print("For n = 4, the number of distinct solutions is:", totalNQueens(4))  # Output: 2
    print("For n = 1, the number of distinct solutions is:", totalNQueens(1))  # Output: 1